### PR TITLE
Create new repository [SCI-1268]

### PR DESCRIPTION
**Pull request by mlorb**
[SCI-1268](https://biosistemika.atlassian.net/browse/SCI-1268)

### CODE REVIEW

File: app/views/repositories/index.html.erb [hunk 1]
```diff
@@ -35,3 +48,6 @@
     <p><%=t 'repositories.index.no_teams.text' %></p>
   </div>
 <% end %>
+
+<%= render partial: "repositories/index/create_new_modal.html.erb" %>
```
Comment (mz3944) [inline app/views/repositories/index.html.erb]: Put this on top, where the other partial is already.

File: app/controllers/repositories_controller.rb [hunk 2]
```diff
@@ -17,4 +57,13 @@ def load_vars
   def check_view_all_permissions
     render_403 unless can_view_team_repositories(@team)
   end
+
+  def check_create_permissions
+    render_403 unless can_create_new_repository(@team) &&
+                      @repositories.count < Constants::REPOSITORIES_LIMIT
```
Comment (mz3944) [inline app/controllers/repositories_controller.rb]: Nojs! :+1:

File: app/views/repositories/index/_create_new_modal.html.erb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        <h4 class="modal-title" id="create-new-modal-label">
+          <%= t("repositories.create.title") %>
+        </h4>
+      </div>
+      <div class="modal-body"></div>
+      <div class="modal-footer">
+        <button type="button" class="btn btn-primary" data-action="submit"><%= t("repositories.create.submit") %></button>
```
Comment (mz3944) [inline app/views/repositories/index/_create_new_modal.html.erb]: Use f.submit instead and you won't need JS code for this.

File: app/assets/javascripts/repositories/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      error: function (error) {
+        // TODO
+      }
+    });
+  });
+
+  submitBtn.on("click", function() {
+    // Submit the form inside modal
+    $(this).closest(".modal").find(".modal-body form").submit();
+  });
```
Comment (mz3944) [inline app/assets/javascripts/repositories/index.js]: Yu don't need this, just use `f.submit`.

File: app/views/repositories/index.html.erb [hunk 5]
```diff
@@ -4,29 +4,42 @@
...
+        <a href="#custom_repo_<%= repo.id %>"
+           data-toggle="tab"
+           aria-controls="custom_repo_<%= repo.id %>"
+           title="<%=repo.name%>"><%= truncate(repo.name, length: Constants::NAME_TRUNCATION_LENGTH) %></a>
+      </li>
+    <% end %>
+    <!-- Add new repository tab -->
+    <li role="presentation"
+      <% unless can_create_new_repository(current_team) &&
+                @repositories.count < Constants::REPOSITORIES_LIMIT %>
```
Comment (mz3944) [inline app/views/repositories/index.html.erb]: more limit check to permissions helper.

File: config/routes.rb [hunk 6]
```diff
@@ -123,7 +123,11 @@
         as: 'file_expired'
 
     resources :teams do
-      resources :repositories, only: [:index]
+      resources :repositories, only: [:index, :create] do
+        collection do
```
Comment (mlorb) [inline config/routes.rb]: I need it.

File: app/views/repositories/index.html.erb (hunk 1 above)
Comment (mlorb) [inline app/views/repositories/index.html.erb]: :+1:

File: app/assets/javascripts/repositories/index.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    $.ajax({
+      url: link.attr("data-url"),
+      type: "GET",
+      dataType: "json",
+      success: function (data) {
+        var modalBody = modal.find(".modal-body");
+        modalBody.html(data.html);
+
+        modalBody.find("form")
+        .on("ajax:success", function(ev2, data2, status2) {
```
Comment (mlorb) [inline app/assets/javascripts/repositories/index.js]: :+1:

File: app/assets/javascripts/repositories/index.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+          // Redirect to index page
+          $(location).attr("href", data2.url);
+        })
+        .on("ajax:error", function(ev2, data2, status2) {
```
Comment (mlorb) [inline app/assets/javascripts/repositories/index.js]: :+1:

File: app/views/repositories/index.html.erb (hunk 5 above)
Comment (mlorb) [inline app/views/repositories/index.html.erb]: :+1:

File: app/views/repositories/index.html.erb [hunk 9]
```diff
@@ -4,29 +4,42 @@
...
+        class="disabled"
+      <% end %>>
+      <a href='#'
```
Comment (mlorb) [inline app/views/repositories/index.html.erb]: no

File: app/controllers/repositories_controller.rb (hunk 2 above)
Comment (mlorb) [inline app/controllers/repositories_controller.rb]: :+1:

File: app/views/repositories/index.html.erb [hunk 10]
```diff
@@ -4,29 +4,42 @@
...
+         data-url="<%= create_new_modal_team_repositories_path %>"
+        <% if can_create_new_repository(current_team) &&
+              @repositories.count < Constants::REPOSITORIES_LIMIT %>
```
Comment (mlorb) [inline app/views/repositories/index.html.erb]: :+1:

File: app/views/repositories/index.html.erb [hunk 11]
```diff
@@ -4,29 +4,42 @@
...
+          data-action='create-new-repository'
```
Comment (mlorb) [inline app/views/repositories/index.html.erb]: :+1:

File: app/controllers/repositories_controller.rb [hunk 12]
```diff
@@ -1,11 +1,51 @@
...
+      team: @team,
+      created_by: current_user
+    )
+    @repository.assign_attributes(create_params)
+
+    respond_to do |format|
+      if @repository.save
+        flash[:success] = t("repositories.create.success_flash", name: @repository.name)
+        format.json {
+          render json: { url: team_repositories_path(@team, create_action: true) },
```
Comment (mlorb) [inline app/controllers/repositories_controller.rb]: not true

File: app/views/repositories/index.html.erb [hunk 13]
```diff
@@ -4,29 +4,42 @@
...
-        <div class="tab-pane tab-pane-settings <%= 'active' if i == 0 %>" id="custom_repo_<%= repo.id %>">
-          <!-- Tab Content -->
-        </div>
-      <% end %>
-    </div>
-  <% else %>
-    <p><em><%=t 'repositories.index.no_repositories' %></em></p>
-  <% end %>
+  <ul class="nav nav-tabs nav-settings" role="tablist" id="tabs">
+    <% active_tab = params[:create_action].present? ? ( @repositories.length - 1 ) : 0 %>
```
Comment (mlorb) [inline app/views/repositories/index.html.erb]: not possible

File: app/assets/javascripts/repositories/index.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+      type: "GET",
+      dataType: "json",
+      success: function (data) {
+        var modalBody = modal.find(".modal-body");
+        modalBody.html(data.html);
+
+        modalBody.find("form")
+        .on("ajax:success", function(ev2, data2, status2) {
+          // Redirect to index page
+          $(location).attr("href", data2.url);
```
Comment (mlorb) [inline app/assets/javascripts/repositories/index.js]: no

File: app/views/repositories/index.html.erb [hunk 15]
```diff
@@ -4,68 +4,79 @@
...
+      </li>
+    <% end %>
+    <!-- Add new repository tab -->
+    <li role="presentation"
+      <% unless can_create_new_repository(current_team) %>
+        class="disabled"
+      <% end %>>
+      <a
+        <% if can_create_new_repository(current_team) %>
+          href="<%= create_new_modal_team_repositories_path %>"
```
Comment (mz3944) [inline app/views/repositories/index.html.erb]: create_modal...*

File: app/controllers/repositories_controller.rb [hunk 16]
```diff
@@ -3,11 +3,48 @@ class RepositoriesController < ApplicationController
...
+      created_by: current_user
+    )
+    @repository.assign_attributes(repository_params)
+
+    respond_to do |format|
+      format.json do
+        if @repository.save
+          flash[:success] = t('repositories.index.modal_create.success_flash',
+                              name: @repository.name)
+          render json: { url: team_repositories_path(repository: @repository) },
```
Comment (mlorb) [inline app/controllers/repositories_controller.rb]: ignoring...

**Pull request merged**